[nccl] Do no skip send/recv 0 byte tensor

xw285cornell opened the pull request:
Summary:
Since NCCL 2.12.10, NCCL supports send/recv 0 byte: https://github.com/NVIDIA/nccl/issues/696. Therefore we don't have to skip.

One issue is that if a rank has 0 bytes to send and 0 bytes to recv, it'll skip send/recv completely. And it'll proceed to the next collective which it can send/recv something, making it confusing to the other ranks. Another solution is to add a barrier but that's very expensive.

Test Plan: will add a unit test

Differential Revision: D46507785



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @anijain2305

malfet commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -850,7 +855,12 @@ void all2all_single_unequal_split(
   for (const auto r : c10::irange(numranks)) {
     // NCCL uses 0 byte message for synchronization
     // Avoid send/recv when message size is zero
+#if defined(NCCL_MAJOR) && \
+    ((NCCL_MAJOR > 2) || ((NCCL_MAJOR == 2) && (NCCL_MINOR > 13)))
+    if (true) {
+#else
```
Consider creating helper function
```
#if defined(NCCL_MAJOR) && \
    ((NCCL_MAJOR > 2) || ((NCCL_MAJOR == 2) && (NCCL_MINOR > 13)))
template<typename T>
constexpr bool _nccl_should_send(C10_UNUSED T _unused_) {
  return foo;
}
#else
template<typename T>
inline bool _nccl_should_send(C10_UNUSED T value) {
  return value != 0;
}
#endif 
```

or just `constexpr bool _nccl_supports_zero_send` and `(_nccl_supports_zero_send || original_cond)` it throughout the code

kwen2501 commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -786,8 +786,8 @@ void all2all_single_equal_split(
-#if defined(NCCL_MAJOR) && (NCCL_MAJOR == 2) && \
-    (NCCL_MAJOR * 10 + NCCL_MINOR) >= 27
+#if defined(NCCL_MAJOR) && \
+    ((NCCL_MAJOR > 2) || ((NCCL_MAJOR == 2) && (NCCL_MINOR > 7)))
```
Is this new condition equivalent to >=27 ? seems to rule out NCCL_MINOR=7.

xw285cornell replied:
ah good catch! Will fix this

xw285cornell committed:
[nccl] Do no skip send/recv 0 byte tensor (#103140)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/103140

Since NCCL 2.12.10, NCCL supports send/recv 0 byte: https://github.com/NVIDIA/nccl/issues/696. Therefore we don't have to skip.

One issue is that if a rank has 0 bytes to send and 0 bytes to recv, it'll skip send/recv completely. And it'll proceed to the next collective which it can send/recv something, making it confusing to the other ranks. Another solution is to add a barrier but that's very expensive.

Example:

rank0: collective1: send/recv all 0s  -> collective2: send 0 and recv non-zero bytes
rankX: collective1: send/recv from all ranks

For rank X, it will indeed skipping send to/recv from rank0. However rank0 is like collective 1 never happens and it goes directly to collect2 where it issues recv non-zero bytes. In that case I think it confuses rankX (even if it doesn't intend to receive from rank0 and it stuck right there - maybe related with group launch?)

Test Plan:
Added a unit test in sigmoid. Maybe we should move it to c10d test?

  buck test mode/opt -c hpc_comms.use_nccl=2.17.1 sigmoid/backend/test_gpu:dist_collectives_test_4gpu
This pass

  buck test mode/opt -c hpc_comms.use_nccl=2.13.4 sigmoid/backend/test_gpu:dist_collectives_test_4gpu
This fails.

Reviewed By: malfet, kwen2501

Differential Revision: D46507785

fbshipit-source-id: 82338bf4467c8abb28b9a677acb77709971c4c83
